core/src/main/scala/org/apache/spark/eventhubs/common/rdd/EventHubsRDD.scala

```diff
@@ -17,139 +17,116 @@
-                                   val nameAndPartition: NameAndPartition,
-                                   val fromOffset: Long,
-                                   val fromSeq: Long,
-                                   val untilSeq: Long,
-                                   val offsetType: EventHubsOffsetType)
-    extends Partition {
-  override def index: Int = sparkPartitionId
-}
-
 private[spark] class EventHubsRDD(sc: SparkContext,
-                                  ehParams: Map[String, Map[String, String]],
-                                  val offsetRanges: List[OffsetRange],
-                                  batchTime: Long,
-                                  offsetParams: ProgressTrackerParams,
-                                  receiverFactory: (Map[String, String]) => Client)
-    extends RDD[EventData](sc, Nil) {
+                                  val ehConf: EventHubsConf,
+                                  val offsetRanges: Array[OffsetRange],
+                                  receiverFactory: (EventHubsConf => Client))
+    extends RDD[EventData](sc, Nil)
+    with Logging
+    with HasOffsetRanges {
+
+  import org.apache.spark.eventhubs.common._
 
   override def getPartitions: Array[Partition] = {
-    offsetRanges.zipWithIndex.map {
-      case (offsetRange, index) =>
-        new EventHubRDDPartition(index,
-                                 offsetRange.nameAndPartition,
-                                 offsetRange.fromOffset,
-                                 offsetRange.fromSeq,
-                                 offsetRange.untilSeq,
-                                 offsetRange.offsetType)
-    }.toArray
+    for {
+      (o, i) <- offsetRanges.zipWithIndex
+    } yield new EventHubsRDDPartition(i, o.nameAndPartition, o.fromSeqNo, o.untilSeqNo)
   }
 
-  private def wrappingReceive(nameAndPartition: NameAndPartition,
-                              client: Client,
-                              expectedEvents: Int,
-                              untilSeqNo: Long): List[EventData] = {
-    val receivedBuffer = new ListBuffer[EventData]
-    while (receivedBuffer.size < expectedEvents) {
-      val receivedEvents = client.receive(expectedEvents - receivedBuffer.size)
-      if (receivedEvents == null) return receivedBuffer.toList // no more messages
-      receivedBuffer ++= receivedEvents
-      if (receivedBuffer.nonEmpty &&
-          receivedBuffer.last.getSystemProperties.getSequenceNumber >= untilSeqNo) {
-        // TODO: What is untilSeqNo set to on first batch when Enqueue Time is used?
-        return receivedBuffer.toList
+  override def count: Long = offsetRanges.map(_.count).sum
+
+  override def isEmpty(): Boolean = count == 0L
+
+  override def take(num: Int): Array[EventData] = {
+    val nonEmptyPartitions =
+      this.partitions.map(_.asInstanceOf[EventHubsRDDPartition]).filter(_.count > 0)
+
+    if (num < 1 || nonEmptyPartitions.isEmpty) {
+      return new Array[EventData](0)
```

Comment:
`Array.empty[EventData]` will skip an allocation

Reply:
Agreed, will change to ```return Array()```.